libexec/rtld-elf/aarch64/reloc.c

```diff
@@ -349,20 +349,39 @@ reloc_plt(Obj_Entry *obj, int flags, RtldLockState *lockstate)
 	    FUNC_PTR_REMOVE_PERMS);
 #endif
 	for (rela = obj->pltrela; rela < relalim; rela++) {
-		Elf_Addr *where;
+		uintptr_t *where;
+#ifdef __CHERI_PURE_CAPABILITY__
+		Elf_Addr *fragment;
+#endif
 
-		where = (Elf_Addr *)(obj->relocbase + rela->r_offset);
+		where = (uintptr_t *)(obj->relocbase + rela->r_offset);
+#ifdef __CHERI_PURE_CAPABILITY__
+		fragment = (Elf_Addr *)where;
+#endif
 
 		switch(ELF_R_TYPE(rela->r_info)) {
 #ifdef __CHERI_PURE_CAPABILITY__
 		case R_MORELLO_JUMP_SLOT:
 			/*
-			 * XXX: This would be far more natural if the linker
-			 * made it an R_MORELLO_RELATIVE-like fragment instead.
-			 * https://git.morello-project.org/morello/llvm-project/-/issues/19
+			 * Old ABI:
+			 *   - Treat as R_AARCH64_JUMP_SLOT
+			 *
+			 * New ABI:
+			 *   - Same representation as R_MORELLO_RELATIVE
+			 *
+			 * Determine which this is based on whether there's
+			 * non-zero metadata next to the address. Remove once
+			 * the new ABI is old enough that we can assume it is
+			 * in use.
 			 */
-			*(uintptr_t *)where = cheri_sealentry(jump_slot_base +
-			    *where);
+			if (fragment[1] == 0)
+				*where = cheri_sealentry(jump_slot_base +
+				    (Elf_Addr)*where);
```

Comment:
Maybe use `fragment[0]` here instead of `(Elf_Addr)*where`?  That would be closer to the previous code that only read an `Elf_Addr` from the existing location rather than a `uintptr_t`